frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx

```diff
@@ -52,510 +52,441 @@ import {
-    if (isNullOrUndefined(value)) {
-      return null
-    }
+export interface Props {
+  disabled: boolean
+  element: NumberInputProto
+  widgetMgr: WidgetStateManager
+  width: number
+  theme: EmotionTheme
+  fragmentId?: string
+}
 
-    const { format, step } = this.props.element
-    let formatString = getNonEmptyString(format)
-
-    /**
-     * In the instance where a step is specified for a float,
-     * and a format is not explicitly provided, we will use the
-     * precision of the step to format the result.
-     */
-    if (formatString == null) {
-      const strStep = step.toString()
-      if (this.isFloatData() && step !== 0 && strStep.includes(".")) {
-        const decimalPlaces = strStep.split(".")[1].length
-        formatString = `%0.${decimalPlaces}f`
+export const NumberInput: React.FC<Props> = ({
+  disabled,
+  element,
+  widgetMgr,
+  width,
+  theme,
+  fragmentId,
+}: Props): ReactElement => {
+  const min = element.hasMin ? element.min : -Infinity
+  const max = element.hasMax ? element.max : +Infinity
+
+  const [step, setStep] = React.useState<number>(getStep(element))
+  const initialValue = getInitialValue({ element, widgetMgr })
+  const [dirty, setDirty] = React.useState(false)
+  const [value, setValue] = React.useState<number | null>(initialValue)
+  const [formattedValue, setFormattedValue] = React.useState<string | null>(
+    formatValue({ value: initialValue, ...element, step })
+  )
+  const [isFocused, setIsFocused] = React.useState(false)
+  const inputRef = React.useRef<HTMLInputElement | HTMLTextAreaElement>(null)
+  const formClearHelper = React.useRef(new FormClearHelper())
+  const id = React.useRef(uniqueId("number_input_"))
+
+  const canDec = canDecrement(value, step, min)
+  const canInc = canIncrement(value, step, max)
+
+  // update the step if the props change
+  React.useEffect(() => {
+    setStep(getStep({ step: element.step, dataType: element.dataType }))
+  }, [element.dataType, element.step])
+
+  const commitValue = React.useCallback(
+    ({ value, source }: { value: number | null; source: Source }) => {
+      console.log({ value })
+      if (notNullOrUndefined(value) && (min > value || value > max)) {
+        inputRef.current?.reportValidity()
+      } else {
+        const newValue = value ?? element.default ?? null
+
+        switch (element.dataType) {
+          case NumberInputProto.DataType.INT:
+            widgetMgr.setIntValue(element, newValue, source, fragmentId)
+            break
+          case NumberInputProto.DataType.FLOAT:
+            widgetMgr.setDoubleValue(element, newValue, source, fragmentId)
+            break
+          default:
+            throw new Error("Invalid data type")
+        }
+
+        setDirty(false)
+        setValue(newValue)
+        setFormattedValue(formatValue({ value: newValue, ...element, step }))
       }
-    }
+    },
+    [min, max, inputRef, element, widgetMgr, fragmentId, step]
```

Comment:
it might be worth to double-check that the `element` here does not change with every re-run of the app. Since it is an object, we might need to do a deep-object check or assign the relevant fields prior to the `useCallback` call and only pass those fields here in the deps array.